### DISCUSSION TITLE
Docs Clarification on `default_factory` dataclass relationship configuration

### DISCUSSION THREAD

**Discussion by PelicanQ** [sqlalchemy/sqlalchemy#10900] [Ideas]
Hi!

I believe some clarification is needed in the beginning of this section: https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#relationship-configuration

When reading this section I am confused, since I am using `MappedAsDataclass` on all my models, and I declare a relationship attribute on a model like: 

```
class BaseModel(MappedAsDataclass, DeclarativeBase):
    pass

class User(BaseModel):
    ... 

class Company(BaseModel):
    ...
    users: Mapped[list[User]] = relationship(init=False)
```

without ever using the `default_factory=list` argument to `relationship()` and it works fine. This section makes it seem like that shouldn't work. Perhaps is the case only for `@reg.mapped_as_dataclass` and not `MappedAsDataclass` mixin?

I am particularly confused by "optional keyword argument" from:

> When specifying a collection-based relationship() as an optional keyword argument, the relationship.default_factory parameter must be passed and it must refer to the collection class that’s to be used.

Since when is `relationship` a keyword argument? And keyword argument to what?

> While the relationship.default_factory can be automatically derived from the given collection class of the relationship() itself, this would break compatibility with dataclasses,

In my case, creating a `Company()` will work fine, so it seems to already derive the default collection to use for `users` as `list`.

This is why I think some clarification would be nice on that section.

**Comment by zzzeek** [comment]
> I am particularly confused by "optional keyword argument" from:

the purpose of dataclasses is to create a class that automatically has an `__init__()` method with arguments on it.  These arguments are derived from the fields.  If I want to make a class called `Blue` with an optional keyword argument `a`, I could write:

```py

@dataclass
class Blue:
    a = dataclasses.field(default=None)

```

that is, using "default=None" makes it an optional argument.

the document referenced is indicating that you can't use "default=[]" for a relationship that refers to a collection, it has to be a factory, `default_factory=list` or similar.

in your case, you are creating a relationship with init=False so you are not indicating this field would be an argument to `__init__` at all.

**Reply by PelicanQ** [reply]
Thanks for answering!

So, if the point of the docs section referenced is to explain that with dataclasses you must use `default_factory` **and not** `default` keyword argument to the `relationship()` to define the default collection to the constructor `__init__`, I think it could be said more clearly. Because this fact, that doing `default=[]` is bad and will mean the same list is fed to every call to `__init__`, I already knew, but it was not what came to mind reading the section.

I still don't understand the "optional keyword argument" part of:

> When specifying a collection-based relationship() as an optional keyword argument, the relationship.default_factory parameter must be passed and it must refer to the collection class that’s to be used.

Like, we are not ever feeding `relationship()` itself as a keyword argument to something. The text sound like `some_func(some_optional_kw=relationship())`. If the sentence is referring to keyword arguments of the models `__init__()`, then it could explicitly say that.

So explicitly saying keyword argument to what, and also clarifying that it is `default_factory` INSTEAD of `default` would make the intended message clear. For instance changing 
> the relationship.default_factory parameter must be passed

to

> the relationship.default_factory parameter must be passed, and not relationship.default

I don't know if my source of confusion is clear here lol

**Reply by zzzeek** [reply]
it's pretty clear (re: confusion as to what the docs state), I agree it's a bad doc as it refers to these concepts kind of obtusely

**Reply by zzzeek** [reply]
so yes we can reword it to be more clear what it's referring to
